Enabling SOF Support For AMD Platform

balakishorepati opened the pull request:
This Patch Series will enable SOF based ACP(Audio Co Processor) firmware solution on Renoir platform.

sofci commented:
Can one of the admins verify this patch?
> reply **test&nbsp;this&nbsp;please** to run this test once

gkbldcig commented:
Can one of the admins verify this patch?

plbossart commented:
test this please

plbossart commented:
@bhiregoudar @sunilkumardommati @ajitkupandey @balakishorepati you should have an invitation in your inbox to join the SOF project. Please accept, otherwise the GitHub and Intel CI will ask for a maintainer to approve tests for each push. Thank you.

lgirdwood commented:
> @bhiregoudar @sunilkumardommati @ajitkupandey @balakishorepati you should have an invitation in your inbox to join the SOF project. Please accept, otherwise the GitHub and Intel CI will ask for a maintainer to approve tests for each push. Thank you.

Welcome aboard guys - btw does AMD have plans for hosting any CI ? 

Fwiw, there is a lot of "vanilla" CI today testing generic functions on GH and also some Intel CI testing code on Intel HW platforms. I would recommend a similar AMD CI when time permits too as this would help us all determine which PRs to merge and which to rework.

plbossart commented on src/platform/amd/renoir/include/platform/chip_registers.h:
```diff
@@ -0,0 +1,571 @@
+	unsigned int	: 29;
+	} bits;
+unsigned int	u32all;
+} acp_dma_cntl_0_t;
```
the indentation seems off in this file?

lyakh replied:
oh... automatically generated? Maybe the script can be improved a bit ;-)

lgirdwood replied:
Looks like this is generated, so it should also come under the same style exception as the Cadence xtensa headers.

plbossart commented on src/platform/amd/renoir/include/platform/fw_scratch_mem.h:
```diff
@@ -0,0 +1,121 @@
+
+#pragma pack(push, 4)
+
+/* MAX number of DMA descripotos */
```
descriptors

plbossart commented on src/platform/amd/renoir/include/platform/lib/dma.h:
```diff
@@ -0,0 +1,30 @@
+#define DMA_ID_DAI		2
+
+#define dma_chan_irq(dma, chan) \
+	dma_irq(dma)
```
single line?

plbossart commented on src/drivers/amd/renoir/acp_dmic_dma.c:
```diff
@@ -0,0 +1,378 @@
+{
+	if (channel->direction == DMA_DIR_DEV_TO_MEM) {
+		*avail = dmic_rngbuff_size/2;
+		*free = dmic_rngbuff_size/2;
```
that seems wrong?

Usually avail+free=buffer size, but why would avail and free be equal?

bhiregoudar replied:
This is at the half buffer water mark, so available and free would be equall.

plbossart replied:
No, this function is supposed to provide the actual number of bytes with data and the number of bytes free.
These numbers will vary dynamically, there is no reason why they should be pegged to the half-buffer value.
This is NOT a watermark.

plbossart replied:
btw avail and free are the same concept in ALSA, see snd_pcm_status()

bhiregoudar replied:
This function provides number of bytes of data available and free in the dai buffer(DSP memory buffer) for the DMIC capture use case, not the data available or free in the host buffer (system memory). 
This data is collected based on the DMIC data copied in DSP memory.

lgirdwood replied:
So SOF uses the same concepts as ALSA for both host side and DSP side DAI endpoint buffers. This is probably working today if your buffer is large enough or your DMIC DMA has divided the buffer into two periods (a ping-pong buffer). You may need to update this in the future to support different use cases.

plbossart commented on src/include/ipc/dai.h:
```diff
@@ -68,6 +68,9 @@ enum sof_ipc_dai_type {
 	SOF_DAI_IMX_ESAI,               /**< i.MX ESAI */
+	SOF_DAI_AMD_BT,			/**< Amd BT */
+	SOF_DAI_AMD_SP,			/**< Amd SP */
+	SOF_DAI_AMD_DMIC		/**< Amd DMIC */
```
ABI change, this should be tagged as such

lgirdwood replied:
@balakishorepati this also means we need the same change in the Linux kernel too. i.e. to align the headers.

balakishorepati replied:
yes same changes has be done in Linux kernel headers.

bhiregoudar replied:
> ABI change, this should be tagged as such
Is there a need to change ABI version ? As per my understanding we don't need to as we didn't change any thing with respect to IPC interface or IPC structures.
Could you also provide more details on tagging the ABI, what needs to be done exactly?

lgirdwood replied:
@bhiregoudar I've added this in the ABI tool, this just helps us align the kernel drivers with the FW. See https://github.com/orgs/thesofproject/projects/2 we need to do the same for your kernel PR. This will mean AMD support will appear with ABI 3.19

lgirdwood replied:
Kernel https://github.com/thesofproject/linux/pull/3054 aligned with this PR for ABI 3.19

plbossart commented on src/platform/amd/renoir/lib/dai.c:
```diff
@@ -54,7 +54,7 @@ static struct dai spdai[] = {
 	}
 };
-
+#ifdef ACP_BT_ENABLE
```
how is this set?

plbossart replied:
should this be part of an earlier patch?

balakishorepati replied:
BT Dai is not the default option on Renoir platforms(HW dependency). 
based on config "CONFIG_AMD_BT" defined in renoir_defconfig, we will enable the ACP_BT_ENABLE macro, which in turn enable the BT Dai initializations.

plbossart commented on src/ipc/dma-copy.c:
```diff
@@ -146,7 +146,11 @@ int dma_copy_new(struct dma_copy *dc)
+	dc->chan = dma_channel_get(dc->dmac, DMA_TRACE_CHANNEL);
+#else
 	dc->chan = dma_channel_get(dc->dmac, 0);
+#endif
```
that should be better abstracted, this doesn't look good to me.

lyakh replied:
maybe we should modify the generic case to remove the hard-coded ß

lgirdwood replied:
@bhiregoudar yep, best thing to do here is to add a CONFIG_TRACE_CHANNEL (done), add more more info in the Kconfig help section and make all platforms use it (all other platforms will be default to 0).

balakishorepati replied:
Addressed the Review Comment. Added generic config to reuse by other platforms as well. 
Added more explanation in Kconfig as well.

lgirdwood replied:
Thanks, but we should change it so everyone uses your new Kconfig option (which defaults to 0 unless set).  So everyone should just use.

```c
dc->chan = dma_channel_get(dc->dmac, CONFIG_DMA_TRACE_CHANNEL);
```

plbossart commented on src/audio/host.c:
```diff
@@ -170,7 +170,11 @@ static void host_update_position(struct comp_dev *dev, uint32_t bytes)
+		hd->local_pos %= hd->host_size;
+#else
 		hd->local_pos = 0;
+#endif
```
we shouldn't have platform-specific ifdefs in common code, that looks likes a missing abstraction.

lgirdwood replied:
@balakishorepati can you provide more context here. How does the host DMA position work on Renoir ?

balakishorepati replied:
It is to update the position on wrap of buffer. It is not necessary that on wrap the buffer position would be zero. At wrap, we might have DMAed some partial data from the beginning of the buffer also. In some cases based on the period size, the last frame may not exactly be the end of the buffer and roll over for some bytes in the beginning. So, the modulus operation will give the actual position info. If the the position info is matching with the buffer size, then the modulus operation will give zero, if there are some bytes consumed from the beginning after wrap, then the modulus operation will give the actual position value.
So Instead of make hard coding it to zero, modules operation will give actual position.

lgirdwood replied:
ok, best to comment this in the source. Fwiw, modulus can be expensive on xtensa, so you may want to work out the position another way if you can.

bhiregoudar replied:
Understand that modulus operation is expensive on Xtensa, but, modulus operation is executed only once at a system memory buffer wrap condition, so believe it won't affect the performance. We will keep an eye on this and look for a better a way.

balakishorepati replied:
Added generic config to reuse by other platforms.

plbossart reviewed (commented):
added a bunch of comments below, most of them cosmetic.

plbossart commented on src/drivers/amd/renoir/acp_dma.c:
```diff
@@ -0,0 +1,621 @@
+	dscr_cnt = sgelems->count;
+	/* Trace uses descriptor from index 7 and other streams use decriptors from 0 */
+	if (channel->index == 7)
+		dscr_strt_idx = 7;
```
I forgot the typo on decriptors

lgirdwood commented on src/drivers/amd/renoir/acp_bt_dai.c:
```diff
@@ -0,0 +1,104 @@
+{
+	//TODO
+	return 0;
+}
```
Not related to this PR, but I think we can probable remove the context save and restore from the general codebase as I don't think they will ever be used. Host drivers and userspace take core of audio context. I will do this work for v1.9

balakishorepati replied:
Addressed the Review Comment. removed the API's .Done

lgirdwood commented on src/drivers/amd/renoir/acp_bt_dai.c:
```diff
@@ -0,0 +1,104 @@
+static inline int btdai_set_config(struct dai *dai, struct ipc_config_dai *common_config,
+				   void *spec_config)
+{
+	//nothing to do on renoir
```
Code style is C style comments i.e. ``` /* comment */ ```

balakishorepati replied:
Addressed the Review Comment. Done

lyakh commented on src/drivers/amd/renoir/interrupt.c:
```diff
@@ -0,0 +1,386 @@
+	acp_dsp_sw_intr_stat_t swintrstat;
+
+	swintrstat = (acp_dsp_sw_intr_stat_t)io_reg_read((PU_REGISTER_BASE +
+							ACP_DSP_SW_INTR_STAT));
```
no need for double parentheses

lyakh commented on src/drivers/amd/renoir/interrupt.c:
```diff
@@ -0,0 +1,386 @@
+	swintrstat = (acp_dsp_sw_intr_stat_t)io_reg_read((PU_REGISTER_BASE +
+							ACP_DSP_SW_INTR_STAT));
+	swintrstat.bits.hosttodsp0intr1stat = INTERRUPT_ENABLE;
+	io_reg_write((PU_REGISTER_BASE + ACP_DSP_SW_INTR_STAT), swintrstat.u32all);
```
in fact there are several locations like this: a comma has the lowest priority in C, so, no need for parentheses around complete function parameters

lyakh commented on src/drivers/amd/renoir/ipc.c:
```diff
@@ -0,0 +1,158 @@
+		0xbb, 0x7e, 0x47, 0x23, 0xf2, 0xe5, 0x73, 0x0c);
+
+struct ipc_data {
+	struct ipc_data_host_buffer dh_buffer;
```
@plbossart a guess: "data host?" :-)

lyakh commented on src/drivers/amd/renoir/ipc.c:
```diff
@@ -0,0 +1,158 @@
+				sof_ipc_host_ack_clear();
+				/* Clear the Host to DSP Status Register */
+				acp_ack_intr_from_host();
+				/*Configures the trigger bit in ACP_DSP_SW_INTR_TRIG register*/
```
yeah, there are multiple such comments in this patch. Also some `//` comments

lyakh commented on src/drivers/amd/renoir/ipc.c:
```diff
@@ -0,0 +1,158 @@
+	acp_sw_intr_trig_t  swintrtrig;
+
+	swintrstat = (acp_dsp_sw_intr_stat_t)io_reg_read((PU_REGISTER_BASE + ACP_DSP_SW_INTR_STAT));
+	status =  (swintrstat.u32all &  HOST_TO_DSP_INTR);
```
too many parentheses in 2 lines above

lgirdwood commented on src/drivers/amd/renoir/acp_bt_dma.c:
```diff
@@ -0,0 +1,446 @@
+				ACP_BT_TX_LINEARPOSITIONCNTR_HIGH);
+		curr_tx_pos = (uint64_t)((tx_high<<32) | tx_low);
+		*free = (curr_tx_pos - prev_tx_pos) > ACP_BT_FIFO_BUFFER_SIZE ?
+			(curr_tx_pos - prev_tx_pos)%ACP_BT_FIFO_BUFFER_SIZE :
```
Need a space before and after the modulus e.g ```) % ACP_```

balakishorepati replied:
Addressed the Review Comment. Done

lgirdwood commented on src/drivers/amd/renoir/acp_bt_dma.c:
```diff
@@ -0,0 +1,446 @@
+				ACP_BT_RX_LINEARPOSITIONCNTR_HIGH);
+		curr_rx_pos = (uint64_t)((rx_high<<32) | rx_low);
+		*free = (curr_rx_pos - prev_rx_pos) > ACP_BT_FIFO_BUFFER_SIZE ?
+			(curr_rx_pos - prev_rx_pos)%ACP_BT_FIFO_BUFFER_SIZE :
```
ditto

balakishorepati replied:
Addressed the Review Comment. Done

lgirdwood commented on src/drivers/amd/renoir/acp_bt_dma.c:
```diff
@@ -0,0 +1,446 @@
+				ACP_BT_TX_LINEARPOSITIONCNTR_LOW);
+		tx_high = (uint32_t)io_reg_read(PU_REGISTER_BASE +
+				ACP_BT_TX_LINEARPOSITIONCNTR_HIGH);
+		curr_tx_pos = (uint64_t)((tx_high<<32) | tx_low);
```
Needs a space e.g. ```tx_high << 32```

balakishorepati replied:
Addressed the Review Comment. Done

lyakh commented on src/platform/amd/renoir/lib/clk.c:
```diff
@@ -0,0 +1,141 @@
+
+void  get_response_from_smu(void)
+{
+	uint32_t retValue = 0;
```
superfluous initialisation

lyakh commented on src/platform/amd/renoir/lib/clk.c:
```diff
@@ -0,0 +1,141 @@
+	acpSrbmCycleSts = (acp_srbm_cycle_sts_t)io_reg_read(PU_REGISTER_BASE + ACP_SRBM_CYCLE_STS);
+	/*Waiting for status to be set to '0'*/
+	while (acpSrbmCycleSts.bits.srbm_clients_sts != 0)
+		acpSrbmCycleSts = (acp_srbm_cycle_sts_t)io_reg_read(PU_REGISTER_BASE + ACP_SRBM_CYCLE_STS);
```
usually such loops have some deterministic abort counter or something similar

lyakh commented on src/platform/amd/renoir/lib/clk.c:
```diff
@@ -0,0 +1,141 @@
+*
+*****************************************************************************/
+int acp_reg_read_via_smn (uint32_t	reg_offset,
+			 uint32_t	reg_value,
```
why do you pass `reg_value` as a parameter? You only use it as a local variable. Was is supposed to be a pointer?

lgirdwood commented on src/drivers/amd/renoir/acp_bt_dma.c:
```diff
@@ -0,0 +1,446 @@
+		return -EINVAL;
+	}
+	return 0;
+}
```
One thing to be wary of here is that I'm assuming your DSP is 32bit and you are reading a 64bit counter here which may overflow during the read (as the 2 * 32 bit reads are not atomic). The SOF Intel CAVS timer driver has the same 64bit readback issue, we need to read it at least twice to make sure the lower word has not overflowed during the read back.

balakishorepati replied:
These are the Registers  read from the HW block .They represent the data consumed.

lyakh commented on src/platform/amd/renoir/lib/clk.c:
```diff
@@ -0,0 +1,141 @@
+	io_reg_write((PU_REGISTER_BASE + ACP_SRBM_CLIENT_BASE_ADDR), smnClientBaseAddr);
+	/*dummy read to the to the address offset 0x3400*/
+	/*(Master config offset to access register outside of ACP)*/
+	reg_value = (uint32_t)io_reg_read(PU_REGISTER_BASE + (ACP_MASTER_REG_ACCESS_ADDRESS + reg_offset - regionStartAddr + ACP_FIRST_REG_OFFSET));
```
`reg_value` is unused

balakishorepati replied:
It is used below lines to send data to caller function.

lyakh commented on src/platform/amd/renoir/lib/clk.c:
```diff
@@ -0,0 +1,141 @@
+		acpSrbmCycleSts = (acp_srbm_cycle_sts_t)io_reg_read(PU_REGISTER_BASE + ACP_SRBM_CYCLE_STS);
+	/*when status is '0' from above read the data from data register*/
+	reg_value = (uint32_t) io_reg_read(PU_REGISTER_BASE + ACP_SRBM_Client_RDDATA);
+	return reg_value;
```
`return io_reg_read(...);`

lyakh commented on src/platform/amd/renoir/lib/clk.c:
```diff
@@ -0,0 +1,141 @@
+	acpSrbmCycleSts = (acp_srbm_cycle_sts_t)io_reg_read(PU_REGISTER_BASE + ACP_SRBM_CYCLE_STS);
+	/*Waiting for status to be set to '0'*/
+	while (acpSrbmCycleSts.bits.srbm_clients_sts != 0)
+		acpSrbmCycleSts = (acp_srbm_cycle_sts_t)io_reg_read(PU_REGISTER_BASE + ACP_SRBM_CYCLE_STS);
```
same with this loop

lyakh commented on src/platform/amd/renoir/lib/clk.c:
```diff
@@ -0,0 +1,141 @@
+		.current_freq_idx = CPU_DEFAULT_IDX,
+		.notification_id = NOTIFIER_ID_CPU_FREQ,
+		.notification_mask = NOTIFIER_TARGET_CORE_MASK(i),
+		.set_freq = NULL,
```
indentation

lgirdwood commented on src/drivers/amd/renoir/acp_dma.c:
```diff
@@ -0,0 +1,621 @@
+#include <stdint.h>
+
+
+#define MAX_STREAMS				8
```
Can we prefix these macros with AMD_ as some are quite generic and may conflict with other generic macros..

balakishorepati replied:
Addressed the Review Comment. Done

lyakh commented on src/drivers/amd/renoir/acp_sp_dma.c:
```diff
@@ -0,0 +1,416 @@
+	uint32_t sp_fifo_addr;
+
+	if (!config->cyclic) {
+		tr_err(&acp_sp_tr, "SPDMA: cyclic configurations are not supported!");
```
same comment: something seems wrong here

lgirdwood commented on src/drivers/amd/renoir/ipc.c:
```diff
@@ -0,0 +1,158 @@
+
+/* 49be8ff3-71a3-4456-bb7e-4723f2e5730c */
+DECLARE_SOF_UUID("ipc-task", ipc_task_uuid, 0x49be8ff3, 0x71a3, 0x4456,
+		0xbb, 0x7e, 0x47, 0x23, 0xf2, 0xe5, 0x73, 0x0c);
```
Only 1 will be built at a time, but I guess you are thinking about bug and error reporting all using the same name string ?
We could have ```renoir-ipc-task```, ```cavs-ipc-task``` and so on ?

lgirdwood commented on src/include/sof/drivers/acp_dai_dma.h:
```diff
@@ -0,0 +1,44 @@
+#define ACP_DMA_BUFFER_ALIGN	64
+#define ACP_DMA_TRANS_SIZE	64
+#define AMD_DAI_DMA_BUFFER_PERIOD_COUNT		2
+#define DRAM_ADDRESS_MASK			0x0FFFFFFF
```
Some of the macros also need a AMD_ or ACP_ prefix as naming is generic and may conflict.

balakishorepati replied:
Addressed the Review Comment. Done

lgirdwood commented on src/platform/amd/renoir/include/platform/chip_registers.h:
```diff
@@ -0,0 +1,571 @@
+
+typedef	union acp_dsp_sw_intr_stat {
+	struct {
+	unsigned int	hosttodsp0intr1stat:1;
```
agree with @plbossart these are quite hard to read. best to use more ```_```

balakishorepati replied:
Addressed the Review Comment. Done

lyakh commented on src/drivers/amd/renoir/acp_dmic_dma.c:
```diff
@@ -0,0 +1,378 @@
+			io_reg_read(PU_REGISTER_BASE + ACP_WOV_PDM_DMA_ENABLE);
+		while (!(uint32_t)pdm_dma_enable.bits.pdm_dma_en_status) {
+			pdm_dma_enable = (acp_wov_pdm_dma_enable_t)
+			io_reg_read(PU_REGISTER_BASE + ACP_WOV_PDM_DMA_ENABLE);
```
safety counter?

balakishorepati replied:
Addressed the Review Comment. Added safety timer counter to avoide infinite while loop.

lyakh commented on src/drivers/amd/renoir/acp_dmic_dma.c:
```diff
@@ -0,0 +1,378 @@
+	/* Chech PDM DMA Status */
+	while ((uint32_t)pdm_dma_enable.bits.pdm_dma_en_status) {
+		pdm_dma_enable = (acp_wov_pdm_dma_enable_t)
+			io_reg_read(PU_REGISTER_BASE + ACP_WOV_PDM_DMA_ENABLE);
```
ditto

balakishorepati replied:
Addressed the Review Comment. Done

lyakh commented on src/drivers/amd/renoir/acp_dmic_dma.c:
```diff
@@ -0,0 +1,378 @@
+	}
+	if (!config->cyclic) {
+		tr_err(&acp_dmic_dma_tr,
+			"DMIC DMA: cyclic configurations are not supported!");
```
another "cyclic"

balakishorepati replied:
Addressed the Review Comment. Done

lgirdwood commented on src/platform/amd/renoir/include/platform/chip_registers.h:
```diff
@@ -0,0 +1,571 @@
+	unsigned int	:1;
+	unsigned int	:17;
+	unsigned int	:1;
+	unsigned int	:1;
```
This file looks copy and pasted from somewhere else, but it look like it can be cleaned up to improve readability. i.e. the bitfield could be ```unsigned int :26;``` instead of above as these look like test or reserved bits that should not be touched by users.

balakishorepati replied:
Addressed the Review Comment. Done

lgirdwood commented on src/platform/amd/renoir/include/platform/drivers/interrupt.h:
```diff
@@ -0,0 +1,173 @@
+
+//==================AMD macros ====================================
+
+#define HOST_TO_DSP_INTR 	0x1
```
If these are AMD macros then best to have an AMD_ prefix.

balakishorepati replied:
Addressed the Review Comment. Done

lyakh commented on src/ipc/ipc3/dai.c:
```diff
@@ -57,6 +57,20 @@ int dai_config_dma_channel(struct comp_dev *dev, void *spec_config)
+					    dd->stream_id);
+		break;
+	case SOF_DAI_AMD_SP:
+		// Select if it is TX or RX channel
```
I don't think we use `//` comments in common code

balakishorepati replied:
Addressed the Review Comment. Done

lgirdwood commented on src/platform/amd/renoir/include/platform/drivers/interrupt.h:
```diff
@@ -0,0 +1,173 @@
+*
+* \param [in]   countValue       Cycle count.
+***************************************************************************/
+static __inline__ void set_ccompare0(unsigned int countValue)
```
I think these are already define in the cadence HAL code, but please check as the cadence versions may be different to yours.

balakishorepati replied:
Addressed the Review Comment. Done

lgirdwood commented on src/platform/amd/renoir/include/platform/fw_scratch_mem.h:
```diff
@@ -0,0 +1,121 @@
+#pragma pack(push, 4)
+
+/* MAX number of DMA descripotos */
+#define	MAX_NUM_DMA_DESC_DSCR	64
```
indentation.

balakishorepati replied:
Addressed the Review Comment. Done

lgirdwood commented on src/platform/amd/renoir/include/platform/fw_scratch_mem.h:
```diff
@@ -0,0 +1,121 @@
+
+typedef struct _acp_atu_grp_pte_ {
+	uint32_t lowPart;
+	uint32_t highPart;
```
snake case.

balakishorepati replied:
Addressed the Review Comment. Done

lgirdwood commented on src/platform/amd/renoir/include/platform/fw_scratch_mem.h:
```diff
@@ -0,0 +1,121 @@
+	uint32_t			reserve[];
+} acp_scratch_reg_config_t;
+
+#pragma pack(pop)
```
We need to use ```__attribute__``` here - See the IPC headers.

balakishorepati replied:
Addressed the Review Comment. Done

lgirdwood commented on src/platform/amd/renoir/include/platform/lib/memory.h:
```diff
@@ -0,0 +1,173 @@
+}
+
+
+static inline void platform_shared_commit(void *ptr, int bytes) { }
```
I think this API has been removed.

balakishorepati replied:
Addressed the Review Comment. Done

lgirdwood commented on src/platform/amd/renoir/lib/clk.c:
```diff
@@ -0,0 +1,141 @@
+
+static SHARED_DATA struct clock_info platform_clocks_info[NUM_CLOCKS];
+
+void  get_response_from_smu(void)
```
static ?

balakishorepati replied:
Addressed the Review Comment. Done

lgirdwood commented on src/platform/amd/renoir/lib/clk.c:
```diff
@@ -0,0 +1,141 @@
+		if (count > 0x007fffff) {
+			count = 0;
+		}
+		count++;
```
Do we need a timeout here to avoid looping forever ? i.e. if the ret value does not change as expected.

lgirdwood commented on src/platform/amd/renoir/lib/clk.c:
```diff
@@ -0,0 +1,141 @@
+
+void  get_response_from_smu(void)
+{
+	uint32_t retValue = 0;
```
snake case.

balakishorepati replied:
Addressed the Review Comment. Done

lgirdwood commented on src/platform/amd/renoir/lib/clk.c:
```diff
@@ -0,0 +1,141 @@
+	acpSrbmCycleSts = (acp_srbm_cycle_sts_t)io_reg_read(PU_REGISTER_BASE + ACP_SRBM_CYCLE_STS);
+	/*Waiting for status to be set to '0'*/
+	while (acpSrbmCycleSts.bits.srbm_clients_sts != 0)
+		acpSrbmCycleSts = (acp_srbm_cycle_sts_t)io_reg_read(PU_REGISTER_BASE + ACP_SRBM_CYCLE_STS);
```
looping forever on HW state change. Needs a timeout and error print.

balakishorepati replied:
Addressed the Review Comment. Done

lgirdwood commented on src/drivers/amd/renoir/acp_sp_dma.c:
```diff
@@ -0,0 +1,416 @@
+		tx_high = (uint32_t)io_reg_read(PU_REGISTER_BASE +
+				ACP_I2S_TX_LINEARPOSITIONCNTR_HIGH);
+		curr_tx_pos = (uint64_t)((tx_high<<32) | tx_low);
+		*free = (curr_tx_pos - prev_tx_pos) > sp_buff_size ? (curr_tx_pos - prev_tx_pos)%sp_buff_size : curr_tx_pos - prev_tx_pos;
```
spaces around the % and do we need to check overflows on 64bit reads from HW.

balakishorepati replied:
Addressed the Review Comment. Done

lgirdwood reviewed (commented):
I think this overall looks fine, but it mostly small things that need fixed like coding style differences, macro prefixes, HW state change loop timeouts.

@balakishorepati do you have a followup patch to add GCC toolchain support in the xtensa-overlay repo. Doing this means that the AMD builds can be added to GH CI to make sure the build is always stable.

balakishorepati commented:
> I think this overall looks fine, but it mostly small things that need fixed like coding style differences, macro prefixes, HW state change loop timeouts.
> 
> @balakishorepati do you have a followup patch to add GCC toolchain support in the xtensa-overlay repo. Doing this means that the AMD builds can be added to GH CI to make sure the build is always stable.

Yes we are planning to add followup patch with addressing all review comments. Also planning to add GCC toolchain support in xtensa-overlay soon.

plbossart commented:
> Yes we are planning to add followup patch with addressing all review comments. Also planning to add GCC toolchain support in xtensa-overlay soon.

May I suggest that the tools are added first? That way everything will be ready to test your PR.

cujomalainey commented:
Please check Codestyle CI and push changes to satisfy checkpatch (note SPDX headers issue can be ignored)

balakishorepati commented on CODEOWNERS:
```diff
@@ -46,6 +47,7 @@ src/drivers/intel/cavs/sue-iomux.c	@lyakh
 src/drivers/intel/haswell/*		@randerwang
 src/drivers/imx/**			@dbaluta
 src/drivers/dw/*			@lyakh
+src/drivers/amd/*			@bhiregoudar @sunilkumardommati
```
Do we need to make separate patch for codeowners ?

dbaluta commented on scripts/xtensa-build-all.sh:
```diff
@@ -6,7 +6,7 @@
 SUPPORTED_PLATFORMS=(byt cht bdw hsw apl skl kbl cnl sue icl jsl \
-                    imx8 imx8x imx8m tgl tgl-h)
+                    imx8 imx8x imx8m tgl tgl-h rn)
 BUILD_ROM=no
```
we should rebase this as we added support for imx8ulp.

balakishorepati replied:
Addressed the Review Comment. Done

lyakh commented:
On Thu, 29 Jul 2021, balakishorepati wrote:

> 
> @balakishorepati commented on this pull request.
> 
> ________________________________________________________________________________________________________________________________________________________________________________________________________________________________________
> 
> In src/platform/amd/renoir/lib/clk.c:
> 
> > +*
> +*****************************************************************************/
> +int acp_reg_read_via_smn (uint32_t	reg_offset,
> +			 uint32_t	reg_value,
> +			 uint32_t	size)
> +{
> +	uint32_t	smnClientBaseAddr    = (reg_offset >> 10);
> +	uint32_t	regionStartAddr = (smnClientBaseAddr << 10);
> +	acp_srbm_cycle_sts_t  acpSrbmCycleSts;
> +	/*Configuring the MP1 Aperture Id in SRB_client_config register*/
> +	io_reg_write((PU_REGISTER_BASE + ACP_SRBM_CLIENT_CONFIG), MP1_APERTURE_ID);
> +	/*Configuring the base address of MP1*/
> +	io_reg_write((PU_REGISTER_BASE + ACP_SRBM_CLIENT_BASE_ADDR), smnClientBaseAddr);
> +	/*dummy read to the to the address offset 0x3400*/
> +	/*(Master config offset to access register outside of ACP)*/
> +	reg_value = (uint32_t)io_reg_read(PU_REGISTER_BASE + (ACP_MASTER_REG_ACCESS_ADDRESS + reg_offset - regionStartAddr + ACP_FIRST_REG_OFFSET));
> 
> It is used below lines to send data to caller function.

Yes, sorry, I see that you return `reg_value`, what I meant is that the 
`reg_value` function parameter is unused. It could be just a local 
variable, and even then you don't really need it: first assignment is 
unused, and second can be a direct `return io_reg_read(...);`
Intel Deutschland GmbH
Registered Address: Am Campeon 10, 85579 Neubiberg, Germany
Tel: +49 89 99 8853-0, www.intel.de <http://www.intel.de>
Managing Directors: Christin Eisenschmid, Sharon Heck, Tiffany Doon Silva  
Chairperson of the Supervisory Board: Nicole Lau
Registered Office: Munich
Commercial Register: Amtsgericht Muenchen HRB 186928

paulstelian97 reviewed (dismissed):
Blocking on the fact that a merge commit exists in this pull request. You should rebase instead.

cujomalainey commented:
@balakishorepati the merge commit is still in the chain 5bead9cbf42652ec89ea4a94f4358adf1e9566b3 can you please rebase it out?

bhiregoudar commented on src/drivers/amd/renoir/acp_bt_dma.c:
```diff
@@ -0,0 +1,437 @@
+	acp_bt_tx_intr_watermark_size_t tx_wtrmrk;
+	acp_bt_rx_intr_watermark_size_t rx_wtrmrk;
+
+	volatile acp_scratch_mem_config_t *pscratch_mem_cfg =
```
FIFO address are configurable and that section of memory is used by multiple H/W blocks and FW, so it is necessary to put volatile there

lgirdwood commented:
@balakishorepati thanks for the update, are you able to reply to the review comments above so that we can mark them as resolved and close them. Thanks !

balakishorepati commented on src/drivers/amd/renoir/acp_dma.c:
```diff
@@ -0,0 +1,621 @@
+	uint32_t dest;
+	uint32_t tail, head;
+
+	acp_cfg_dma_descriptor_t psrc_dscr[8];
```
Addressed the Review Comment.

balakishorepati commented on src/drivers/amd/renoir/acp_dma.c:
```diff
@@ -0,0 +1,621 @@
+	atomic_add(&dma->num_channels_busy, 1);
+	channel->status = COMP_STATE_READY;
+	spin_unlock_irq(&dma->lock, flags);
+	/* reset read and write pinters */
```
Addressed the Review Comment. Done

balakishorepati commented on src/drivers/amd/renoir/acp_dma.c:
```diff
@@ -0,0 +1,621 @@
+	acp_dma_cntl_0_t	 dma_cntl;
+	acp_dma_ch_sts_t	 ch_sts;
+	uint32_t dmach_mask	= 0;
+	uint32_t delay_time	= 10000;
```
it is delay_counter. Rename to proper name.

balakishorepati commented on src/drivers/amd/renoir/acp_dma.c:
```diff
@@ -0,0 +1,621 @@
+	/* set dmachrun bit to start the transfer*/
+	dma_chan_reg_write(channel, ACP_DMA_CNTL_0, dma_cntl.u32all);
+
+	/* poll for the status bit - to finish the dma transfer then inititate call back function*/
```
Addressed the Review Comment. Done

balakishorepati commented on src/drivers/amd/renoir/acp_bt_dma.c:
```diff
@@ -0,0 +1,446 @@
+		return -EINVAL;
+	}
+	if (!config->cyclic) {
+		tr_err(&acp_bt_dma_tr, "BT_DMA: cyclic configurations are not supported!");
```
Addressed the Review Comment. Done

balakishorepati commented on src/drivers/amd/renoir/acp_bt_dma.c:
```diff
@@ -0,0 +1,446 @@
+				ACP_BT_TX_LINEARPOSITIONCNTR_HIGH);
+		curr_tx_pos = (uint64_t)((tx_high<<32) | tx_low);
+		*free = (curr_tx_pos - prev_tx_pos) > ACP_BT_FIFO_BUFFER_SIZE ?
+			(curr_tx_pos - prev_tx_pos)%ACP_BT_FIFO_BUFFER_SIZE :
```
Addressed the Review Comment. Done

balakishorepati commented on src/drivers/amd/renoir/acp_dma.c:
```diff
@@ -0,0 +1,621 @@
+#include <sof/audio/component.h>
+#include<platform/chip_registers.h>
+#include<platform/fw_scratch_mem.h>
+#include<platform/chip_offset_byte.h>
```
Addressed the Review Comment. Done

balakishorepati commented on src/drivers/amd/renoir/acp_dma.c:
```diff
@@ -0,0 +1,621 @@
+{
+	acp_dma_cntl_0_t	 dma_cntl;
+	acp_dma_ch_sts_t	 ch_sts;
+	uint32_t dmach_mask	= 0;
```
Addressed the Review Comment. Done

balakishorepati commented on src/drivers/amd/renoir/acp_dma.c:
```diff
@@ -0,0 +1,621 @@
+	switch (cmd) {
+	case DMA_IRQ_STATUS_GET:
+		status =  (dma_reg_read(channel->dma, ACP_DSP0_INTR_STAT) & 0xFF);
+		return (status & (1<<channel->index));
```
Addressed the Review Comment. Done

balakishorepati commented on src/platform/amd/renoir/include/platform/chip_offset_byte.h:
```diff
@@ -0,0 +1,122 @@
+#define ACP_SRBM_CYCLE_STS                            0x1241BC4
+#define ACP_SRBM_CLIENT_CONFIG                        0x1241BC8
+
+// Registers from ACP_AUDIO_BUFFERS block
```
Addressed the Review Comment. Done

balakishorepati commented on src/platform/amd/renoir/include/platform/fw_scratch_mem.h:
```diff
@@ -0,0 +1,117 @@
+
+typedef struct  acp_scratch_memory_config {
+	/* ACP out box buffer */
+	uint8_t				acp_outbox_buffer[512];
```
Addressed the Review Comment. Done

balakishorepati commented on src/audio/Kconfig:
```diff
@@ -473,4 +473,15 @@ config PCM_CONVERTER_FORMAT_CONVERT_HIFI3
+	bool "Buffer wrapping"
+	default n
+	help
+	  Use for buffer wrapping to get actual stream position.
```
Addressed the Review Comment. Done

balakishorepati commented on src/include/kernel/abi.h:
```diff
@@ -29,7 +29,7 @@
 /** \brief SOF ABI version major, minor and patch numbers */
 #define SOF_ABI_MAJOR 3
-#define SOF_ABI_MINOR 19
+#define SOF_ABI_MINOR 20
```
Addressed the Review Comment. Done

lgirdwood reviewed (commented):
Just a few opens.
